elements/dash-tree/dash-tree.js

```diff
@@ -0,0 +1,436 @@
+      while (treeItem && treeItem.getAttribute('role') !== 'treeitem') {
+        treeItem = treeItem.parentElement;
+      }
+
+      this._focusTreeItem(treeItem);
+      this._toggleTreeItem();
+      this._selectTreeItem();
+    }
+
+    /**
+     * Return the current `aria-activedescendant` if there is one. Otherwise,
+     * return null.
+     */
+    _currentTreeItem() {
+      const activedescendant = this.getAttribute('aria-activedescendant');
+      if (activedescendant) {
+        return this.querySelector(`#${activedescendant}`);
+      }
+
+      return null;
+    }
+
+    /**
+     * Attempt to find the next `DashTreeItem` in the list. If one exists,
+     * focus it. Otherwise just ignore the command.
+     */
+    _focusNextTreeItem() {
+      const treeItems = this._allTreeItems();
+      const currentItem = this._currentTreeItem();
+      const idx = treeItems.lastIndexOf(currentItem) + 1;
+      if (idx < treeItems.length) this._focusTreeItem(treeItems[idx]);
+    }
+
+    /**
+     * Attempt to find the previous `DashTreeItem` in the list. If one exists,
+     * focus it. Otherwise just ignore the command.
+     */
+    _focusPrevTreeItem() {
+      const treeItems = this._allTreeItems();
+      const currentItem = this._currentTreeItem();
+      const idx = treeItems.lastIndexOf(currentItem) - 1;
+      if (idx >= 0) this._focusTreeItem(treeItems[idx]);
+    }
+
+    /**
+     * Focus the first `DashTreeItem` in the tree. Useful for when the user
+     * presses the [home] key.
+     */
+    _focusFirstTreeItem() {
+      this._focusTreeItem(this.querySelector('dash-tree-item:first-of-type'));
+    }
+
+    /**
+     * Focus the last `DashTreeItem` in the tree. Useful for when the user
+     * presses the [end] key.
+     */
+    _focusLastTreeItem() {
+      const treeItems = this._allTreeItems();
+      this._focusTreeItem(treeItems[treeItems.length - 1]);
+    }
+
+    /**
+     * Mark the passed in element as the new `aria-activedescendant` and give
+     * it an `.active` class for easy styling.
+     */
+    _focusTreeItem(treeItem) {
+      this.setAttribute('aria-activedescendant', treeItem.id);
+
+      // There can be only one active item at a time.
+      // Find any previous active item and remove its active class.
+      const activeItem = this.querySelector('.active');
+      if (activeItem)
+        activeItem.classList.remove('active');
+
+      treeItem.classList.add('active');
+    }
+
+    /**
+     * Flip the `DashTreeItem` between open and closed states.
+     */
+    _toggleTreeItem() {
+      const treeItem = this._currentTreeItem();
+      if (isExpanded(treeItem)) {
+        this._collapseTreeItem();
+      } else {
+        this._expandTreeItem();
+      }
+    }
+
+    /**
+     * If focus is on an open node, closes the node.
+     * If focus is on a child node that is also either an end node or a closed
+     * node, move focus to its parent node.
+     */
+    _collapseTreeItem() {
```

Comment:
I’d either call this `_collapseCurrentTreeItem()` or – and I prefer that – have this take a parameter.